gui: gtk containers rehauling (and more) [READY TO MERGE]

12345ieee opened the pull request:
This is the branch I was talking about in https://github.com/scanmem/scanmem/issues/167#issuecomment-225192309 .

The current state is ~~a disaster, but I opened a PR so I can reference this changeset elsewhere in the issues and so that people can test the final result, if they want.~~ ready to merge.

Once done it will fix #167 and #74 (which was already fixed in #171 if that's merged before this)

It includes #168 and #171 .

Blocks #174 and will probably include the fix for that, too, in the end.

sriemer commented:
I get lots of warnings on openSUSE Leap 42.1 with GTK 3.16.

````
/usr/share/gameconqueror/GameConqueror.py:111: Warning: The property GtkWidget:margin-left is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(os.path.join(WORK_DIR, 'GameConqueror.xml'))
/usr/share/gameconqueror/GameConqueror.py:111: Warning: The property GtkWidget:margin-right is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(os.path.join(WORK_DIR, 'GameConqueror.xml'))
/usr/share/gameconqueror/GameConqueror.py:111: Warning: The property GtkAlignment:bottom-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(os.path.join(WORK_DIR, 'GameConqueror.xml'))
/usr/share/gameconqueror/GameConqueror.py:111: Warning: The property GtkAlignment:left-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(os.path.join(WORK_DIR, 'GameConqueror.xml'))
/usr/share/gameconqueror/GameConqueror.py:111: Warning: The property GtkAlignment:right-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.
  self.builder.add_from_file(os.path.join(WORK_DIR, 'GameConqueror.xml'))
(GameConqueror.py:10281): Gtk-WARNING **: Theme directory actions/48 of theme breeze has no size field
(GameConqueror.py:10281): Gtk-WARNING **: Theme directory categories/16 of theme breeze has no size field
(GameConqueror.py:10281): Gtk-WARNING **: Theme directory categories/22 of theme breeze has no size field
(GameConqueror.py:10281): Gtk-WARNING **: Theme directory categories/48 of theme breeze has no size field
(GameConqueror.py:10281): Gtk-WARNING **: Theme directory devices/48 of theme breeze has no size field
(GameConqueror.py:10281): Gtk-WARNING **: Theme directory  of theme oxygen has no size field
(GameConqueror.py:10281): Gtk-WARNING **: Error loading image 'file:///usr/share/themes/Breeze/gtk-3.0/assets/scrollbar-slider-overlay-horizontal.png': Error opening file: No such file or directory
````

sriemer commented:
The GtkAlignment warnings can't be fixed for now as that would mean losing GTK 3.10 support. The fix requires at least GTK 3.12. But the theme warnings look strange to me.

sriemer commented:
Hitting this as well, when putting the window too small:
````
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug
````

12345ieee commented:
No matter what I do, I see none of these warnings, I'll see if I can increase the warning output.

Do you still have the resize bug here?

sriemer commented:
I would prefer if growing the size would also enlarge the list of matches a bit so that "Offset" and "Region Type" become visible.

sriemer commented:
I've never seen the resize bug as described in #74. I rather see different resize bugs.

12345ieee commented:
> I would prefer if growing the size would also enlarge the list of matches a bit so that "Offset" and "Region Type" become visible.

Yes, that was in the pipeline, same thing with the bottom panel and vertical resizing.

> I've never seen the resize bug as described in #74. I rather see different resize bugs.

You lost me, I used to have exactly the bug in the #74 video.

12345ieee commented:
About the targeted Gtk release, before this patch the gui used elements from the 2.x, deprecated already in 3.0.

I removed all of them, then I had to bump requirements to 3.2 for the placeholder_text property I added.
For reference, Debian Wheezy has gtk 3.4

12345ieee commented:
The breeze theme warnings are the theme's fault.
See: https://bugs.kde.org/show_bug.cgi?id=362041

I installed the breeze theme and indeed I see those warnings.
The theme I normally use (adwaita) works fine.

12345ieee commented:
I've changed all the padding properties in margin properties, so it will be easy to change all of them when you'll want to drop support for 3.10 (as margin-start and the like is from 3.12, as @sriemer was saying).

12345ieee commented:
After the last commit (bf7d0fb) things should also resize correctly.
Have a look, @sriemer .

sriemer commented:
Yep, looks good and works. Thanks! Only reducing size at the bottom to the minimum still triggers the bug messages.

12345ieee commented:
Try now.

sriemer commented:
Nope, not fixed. As soon as the headline of the cheatlist treeview disappears, the bug messages appear.

12345ieee commented:
I can't do that.
No matter what I do, the treeview can't get smaller than header + ~2 entries.
I'd write it off as a gtk oddity.

sriemer commented:
Thanks, yep might be an openSUSE bug. I'll try on Ubuntu 14.04 trusty later on.

12345ieee commented:
Let's start with the first block, @sriemer .
I ask to merge up to 3e213a9 .

This supersedes #168 (which I'll close if this gets merged) and does:
* fix process labels if cannot read the process maps
* fix typos
* implement some command line options for GC

12345ieee commented:
It's done, @sriemer please review the whole ordeal.

Especially towards the end some unrelated changes crept in, but it was easier to keep building on this rather than entering a rebase hell.

The commit that does ~90% of the work is 94bcf627ed0daa407217223e7e1b717a087cbe8f , I'm sorry but I cannot split it meaningfully, as the 15 commits I squashed together to make it were all broken transient states.

Fixes #167 , #74 and #174 .
Supersedes #168 and #171 .

12345ieee commented:
One more commit, to update the manpage for the new cmdline options.

12345ieee commented:
One more, added i18n support for the lines I added and updated .po's.

sriemer commented on gui/GameConqueror.py:
```diff
@@ -809,8 +809,9 @@ def select_process(self, pid, process_name):
+            self.process_label.set_property('tooltip-text', process_name)
 
         self.command_lock.acquire()
         self.backend.send_command('pid %d' % (pid,))
```
Please write your current state description in the commit description in present. You're writing it for the maintainer and before merging, it is not yet fixed. I've changed the description and merged this. Thanks!

sriemer replied:
Also American English is preferred.

12345ieee replied:
Sorry, in this PR this was the only one with the past tense description.

12345ieee replied:
If I have to be honest, I have no clue which English is British and which is American.
Using emacs to write commit messages does not help.

Tell me and I'll fix it.

sriemer commented on gui/GameConqueror.py:
```diff
@@ -1040,9 +1039,26 @@ def check_backend_version(self):
+    if (args.pid is not None) :
+        process_name = os.popen('ps -p ' + str(args.pid) + ' -o command=').read().strip()
+        gc_instance.select_process(args.pid, process_name)
+    
```
Introduces a whitespace error here.

12345ieee replied:
You mean that you want all your empty lines with no starting spaces?
The whole file is inconsistent on this, I thought it didn't matter.
A later commit actually fixes it.

sriemer commented on gui/GameConqueror.py:
```diff
@@ -23,12 +23,11 @@
 import tempfile
 import platform
 import threading
-import time
```
Dropping `re` and `time` imports is not documented. Are you 100% sure that none of this functionality is included?

12345ieee replied:
My fault!
Yes, my linter complains and searching for `time.` yields no results.

sriemer commented on gui/GameConqueror.py:
```diff
@@ -1040,9 +1039,26 @@ def check_backend_version(self):
+    parser = argparse.ArgumentParser(prog='GameConqueror',
+                                     description="A GUI for scanmem, a game hacking tool"
+
+                                    )
```
Two empty lines here. Such stuff should be avoided.

12345ieee replied:
I introduced this by error when rebasing, see 1866e11f522eb2de002eca6e54543bd04248347c .

12345ieee commented:
So, you want me to fix the messages and update?

sriemer commented on gui/GameConqueror.py:
```diff
@@ -1040,9 +1039,26 @@ def check_backend_version(self):
+                                     description="A GUI for scanmem, a game hacking tool"
+
+                                    )
+    parser.add_argument('-v', '--version', action='version', version='%(prog)s ' + VERSION)
```
It doesn't make much sense to require root privileges to print a version number or the help text. So maybe also the gameconqueror script should be modified. At least openSUSE won't require root privileges at all as they don't have the YAMA security module and don't want to set the `ptrace_scope` to 1.

12345ieee replied:
I know.
The "best" way would be to slot the `-h` and `-v` handling in the bash script and suppress them in the .py , but I was not willing to look up how to parse cmdline args from bash.

You can put it on a TODO, but IMHO it's still an improvement from what was before.

These cmdline arguments are more for fast testing (`python GameConqueror.py $(pidof geany) -s'123'` is my standard invocation) than really for end users.

sriemer commented on gui/GameConqueror.py:
```diff
@@ -1043,8 +1043,8 @@ def check_backend_version(self):
-                                     description="A GUI for scanmem, a game hacking tool"
-
+                                     description="A GUI for scanmem, a game hacking tool",
+                                     epilog='Report bugs to <%s>.' %(PACKAGE_BUGREPORT,)
```
Is trivial enough to be merged with the previous commit.

sriemer replied:
`epilog='Report bugs to ' + PACKAGE_BUGREPORT + '.'` is better readable.

12345ieee replied:
I modelled it on scanmem's help format, you'll want to change it over there, too.

sriemer commented:
Nope, info for next time. First part merged.

sriemer commented on gui/GameConqueror.xml:
```diff
@@ -15,682 +15,9 @@ Special thanks: Igor Gnatenko, Andrea Stacchiotti, Mattias Muenster, Zhang Zhe</
-&lt;http://www.gnu.org/philosophy/why-not-lgpl.html&gt;.</property>
     <property name="authors">Wang Lu &lt;coolwanglu(a)gmail.com&gt;</property>
     <property name="logo">GameConqueror_128x128.png</property>
+    <property name="license_type">gpl-3-0</property>
```
Merged, thanks!

sriemer commented on gui/GameConqueror.xml:
```diff
@@ -866,97 +953,11 @@ Special thanks: Igor Gnatenko, Andrea Stacchiotti, Mattias Muenster, Zhang Zhe</
-    <property name="visible">True</property>
-    <property name="can_focus">False</property>
-    <property name="icon_name">computer</property>
-  </object>
```
Reviewed by doing the same with glade 3.18. Same diff. Also checked manually. Merged. Thanks!

12345ieee commented:
@sriemer You keep commenting on commits that are not actually in the PR, but are functionally identical to some that are.

You may want to refresh your browser, I've probably rebased them since you opened the page.

12345ieee commented:
I made a change to the commit 'gui: Set the correct hint_type for each subwindow' (now 79feb83).
Everything after that was rebased over the new commit.

sriemer commented:
@12345ieee I've almost finished merging. Thanks for your great work!
Please don't put any new patches to your `gui` branch. Rather rebase to the upstream `master` branch then and proceed with a new pull request.

12345ieee commented:
I'm sorry for tacking the 2 extra commits at the end, but they were closely related to this PR, essentially fixing up some mistakes I made in the big rehauling commit.

I am ~20 commits ahead of you, I'll open another PR (after rebasing) for a series of minor changes.
Please merge 047e672 and 77c0fe5 without a separate PR, they are very minimal changes.

sriemer commented:
@12345ieee wrote:
"That's exactly what I did in my diff, @sriemer (I just used another similar property).
If you want a number, on my system (gnome) it feels around 50 would replicate the default behaviour."

@12345ieee: Where is your commit for this?
It is not part of this PR.

12345ieee commented:
Agh, misunderstanding.
I meant the diff I put in the comment before yours: https://github.com/scanmem/scanmem/pull/176#commitcomment-18116361

sriemer commented:
Closing PR as patches were merged or moved to different PR. Thanks!
